benchmark/benchmark.py

```diff
@@ -0,0 +1,209 @@
+        """Begin building Report.MD file
+
+        :return: None
+        """
+        with open("../outputs/report.md", "w") as f:
+            f.write("# The Eyecite Report :eye:\n\n")
+            f.write("\n\nGains and Losses\n")
+            f.write("---------\n")
+            f.write(
+                f"There were {len(self.gains)} gains and "
+                f"{len(self.losses)} losses.\n"
+            )
+            f.write("\n<details>\n")
+            f.write("<summary>Click here to see details.</summary>\n\n")
+
+        # Add markdown report file outputs
+        df = pd.read_csv("../outputs/output.csv")
+
+        with open("../outputs/report.md", "a") as md:
+            if df.__len__() > MAX_ROWS_IN_MD:
+                with open("outputs/report.md", "a+") as f:
+                    f.write(
+                        f"There were {df.__len__()} changes so we are only "
+                        f"displaying the first 50. You can review the \n"
+                        f"entire list by downloading the output.csv "
+                        f"file linked above.\n\n"
+                    )
+
+                df[:MAX_ROWS_IN_MD].to_markdown(buf=md)
+            else:
+                df.to_markdown(buf=md)
+
+        # Remove NAN from file to make it look cleaner
+        with open("../outputs/report.md", "r+") as f:
+            file = f.read()
+            file = re.sub("nan", "   ", file)
+            f.seek(0)
+            f.write(file)
+            f.truncate()
+
+        with open("../outputs/report.md", "a+") as f:
+            f.write("\n\n</details>\n")
+
+        # Add header for time chart for PR comment
+        with open("../outputs/report.md", "a") as f:
+            f.write("\n\nTime Chart\n")
+            f.write("---------\n")
+
+    def generate_time_chart(self) -> None:
+        """Generate time chart showing speed across branches
+
+        return: None
+        """
+
+        main = pd.read_csv(fp_main)
+        branch = pd.read_csv(fp_branch)
+
+        main.columns = main.columns.str.replace("Total", "Total Main")
+        branch.columns = branch.columns.str.replace("Total", "Total Branch")
+
+        df = pd.merge_asof(main, branch, on="Time")
+        df.plot(kind="line", x="Time", y=["Total Main", "Total Branch"])
+
+        plt.ylabel("# Cites Found ", rotation="vertical")
+        plt.savefig("../outputs/time-comparison.png")
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--branch", action="store_true")
+    parser.add_argument("--chart", action="store_true")
+    args = parser.parse_args()
+
+    benchmark = Benchmark()
+    if args.chart:
+        # Process the report
+        benchmark.compare_dataframes()
+
+        # Write Report.MD file
+        benchmark.write_report()
+
+        # Generate time chart
+        benchmark.generate_time_chart()
+    else:
+        # Generate data comparison
+        benchmark.generate_branch_report(branch=args.branch)
```

Comment:
Looking at this a bit, I wonder if it'd make sense to always require a `--ref` parameter. So, right now the approach is:

 - Check out the current ref
 - Clone main
 - Copy benchmark.py from the current ref to main
 - Run `benchmark.py --branch`
 - Check out the current ref
 - Run `benchmark.py`
 - Run `benchmark.py --chart`

(Or something close to that.)

Would it be impractical to have it change to simply:

 - Run `benchmark.py --ref $SOME_SHORT_HASH`
 - Run `benchmark.py --ref $ANOTHER_SHORT_HASH`
 - Run `benchmark.py --generate-report $SOME_SHORT_HASH $ANOTHER_SHORT_HASH`

And in the `artifacts` branch, instead of storing `main.csv` and `branch.csv` it could store $SHORT_HASH.csv, 

The one tricky thing about this is it moves git stuff into the Python script, but maybe it's reasonable to have that here considering the goal is to compare two hashes?

Reply:
im also working to integrate the reporters-db trigger to run the main branch against itself with a reporters-db@HASH install separating the two items.  I think it should be possible to combine them into one test file.

Reply:
let me chew on your suggestion.  thats more or less how I had it - or partially had it before - but with branch names... but I realized I could just drop them ... but I do like the idea of storing the hashes, which would let me check if we have that hash output to speed up the entire test by 50% or so